tools/gfx/persistent-shader-cache.h

```diff
@@ -36,38 +51,50 @@ class PersistentShaderCache : public RefObject
 
     // Update the contents hash for the specified entry in the cache and update the
     // corresponding file on disk.
-    void updateEntry(
-        LinkedNode<ShaderCacheEntry>* entryNode,
-        const slang::Digest& dependencyDigest,
-        const slang::Digest& contentsDigest,
-        ISlangBlob* updatedCode);
+    void updateEntry(const slang::Digest& dependencyDigest, const slang::Digest& contentsDigest, ISlangBlob* updatedCode);
 
 private:
     // Load a previous cache index saved to disk. If not found, create a new cache index
     // and save it to disk as filename.
     void loadCacheFromFile();
 
-    // Update the cache index on disk. This should be called any time an entry changes.
-    void saveCacheToFile();
-
     // Delete the last entry (the least recently used) from entries, remove its key/value pair
-    // from keyToEntry, and remove the corresponding file on disk. This should only be called
-    // by addEntry() when the cache reaches maximum capacity.
-    void deleteLRUEntry();
+    // from keyToEntry, and remove the corresponding file on disk. Returns the index in 'entries'
+    // of the removed entry so addEntry() can overwrite the corresponding entry in 'entries'
+    // with the new entry. This should only be called by addEntry() when the cache reaches maximum capacity.
+    Index deleteLRUEntry();
+
+    // These functions perform the same tasks as their counterparts defined above. However, without access
+    // to a physical file path, in-memory file systems cannot leverage file streams and need to fall back
+    // on a different implementation.
+    void loadCacheFromMemory();
+    ShaderCacheEntry* findEntryInMemory(const slang::Digest& key, ISlangBlob** outCompiledCode);
+    void addEntryToMemory(const slang::Digest& dependencyDigest, const slang::Digest& contentsDigest, ISlangBlob* compiledCode);
```

Comment:
We should find a way to unify most of the logic between os file system path and memory file system path.
I think we define wrappers like  `seekInFile()` and `writeBytesToFile` to intercept the `fseek` `fwrite` calls that modifies a real file for the OS path and just do in-place modification in memory for the memory path.

Reply:
Looking at this again, only loading the cache from file has significant differences in logic due to their different sources. The memory path just needs the extra save call on exit and otherwise shares basically all of its logic with the OS path, which has the extra file stream read/write. Think just wrapping the file stream I/O in an if check for an on-disk file system is sufficient.